### PR TITLE
Bv legend

### PR DESCRIPTION
**Pull request by EbruAyyorgun**
## PR Summary

The Bivariate Legend is a new feature where users can map their data based on 2 variables instead of just 1 variable using colorbars. It can be thought of as a 2D Colorbar. This allows for users to visualize new correlations and extract more meaning from their datasets that previously couldn't be seen because users are able to visualize more variables and factors at one time. The user passes in 2 lists of data that must be the same size. Then, each element of the lists are mapped to their individual colormaps and these colors are blended to result in a bivarate colormap of that single data point. 

The result is a list of rgb colors corresponding to each data point (i.e., combination of the 5th element in the 1st list and the 5th element in the 2nd list produces the 5th rgb color in the resulting list, which represents where the 5th data point was mapped to in the 2D Colorbar). This list of rgb colors can then be passed into the color parameter of any matplotlib plot (see example below)

For more information on bivariate legends, their functions, and their usages in various applications, please refer to the links below. 
https://www.joshuastevens.net/cartography/make-a-bivariate-choropleth-map/ 
https://www.esri.com/arcgis-blog/products/arcgis-living-atlas/mapping/multivariate-maps-what-are-they-and-how-can-i-make-them-in-arcgis/ 

Here's an example usage and resulting image:
`    

    import matplotlib
    from matplotlib import pyplot as plt
    import pandas as pd
    import math
    import numpy as np
    import matplotlib.colors as mcolors

    from sklearn.datasets import load_iris
    iris = load_iris()
    df = pd.DataFrame(data=iris.data)
    df.columns = iris.feature_names

    fig = plt.figure()
    ax = fig.add_subplot(projection='3d')
    norm = matplotlib.colors.PowerNorm(gamma=0.5, vmin=0.0, vmax=1.0)
    blegend = plt.bivariate_legend(
        df['petal length (cm)'], df['petal width (cm)'], d1num_bins=4, d2num_bins=4)
    scatter = ax.scatter(df['sepal length (cm)'], df['sepal width (cm)'],
                        df['petal length (cm)'], c=blegend.get_mapped_colors() , alpha=1)
    plt.show()
`

image

This feature has been useful in ongoing projects involving interpreting high-dimensional RNA-seq data, and it would likely be useful in many other data science projects. 

To accomplish this, I created a new class found in bivariate_legend.py, which can be called from the bivariate_legend function in pyplot.py, which calls the same function in figure.py.

I’ve also included many other customization parameters, including binning the data (seen in the above image where there are boxes that the points fall in to make it easier to interpret with fewer colors), similar extend capabilities as Colorbar, normalization, multiple options for choosing the colormap and colors gradient types, and much more (see docstring in bivariate_legend.py).

Thanks so much for your considerations.

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by jklymak** [comment]
This seems like it would be fine as a stand-alone package rather than needing to be part of Matplotlib core.  

Note that we have had a few attempts at bivariate colormaps over the years, to varying success.  

I'm extremely leery about adding this to the main library.  The bivariate legend proposed here does three things - it bins the data (!) for colormapping purposes, it provides a colormapping, and it makes a bivariate color cube.  That colormapping, so far as I can tell, is only suitable for `scatter` - does it work for anything else where you may want bivariate colors? eg `pcolormesh`, or `contour`?  

My knee-jerk reaction would be that this would best start its life as a third-party package, at least until the API boundaries are worked out.

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -1163,6 +1164,31 @@ def colorbar(
         self.stale = True
         return cb
 
+    @docstring.dedent_interpd
+    def bivariate_legend(self, *args, ax=None, legend_padding=None,
+                         subplot_adjust=None, **kwargs):
+        if (subplot_adjust is None):
```
Comment (jklymak) [inline lib/matplotlib/figure.py:1170]: No matter where this lands, I think that the legend itself should be passed an axes argument - its really to big to be a decoration on an existing axes.  Even if the axes is created with `inset_axes`.    `subplots_adjust` is incompatible with constrained layout, and should rarely be necessary.
